lib/client/Client.js

```diff
@@ -120,29 +118,53 @@ export default class Client {
         clientName: Constants.CLIENT_NAME,
         clientVersion: Constants.VERSION
       })
-      .then(data => {
-        authContinuation = new AuthContinuation(data);
+      .then(data => this._authenticateResponse(data, continuationCallback));
+  }
 
-        return authContinuation;
-      })
-      .then(authContinuation => continuationCallback(authContinuation))
-      .then(continueData => {
-        if (authContinuation.methods.indexOf(continueData.method) < 0) {
-          throw new Error('The "' + continueData.method + '" authentication protocol is not supported by the server.');
-        }
-        let param = {
-          token: authContinuation.continuationToken,
-          method: continueData.method
-        };
+  /**
+   * Sub-routine handling JMAP server responses on authentication requests
+   *
+   * Depending on the server's response, this either (recursively) executes the second
+   * authentication step by calling the provided continuationCallback or resolves on
+   * successfully completed authentication.
+   *
+   * @param data {Object} The JMAP response data
+   * @param continuationCallback {Function} The callback function initially passed to the authenticate() method
+   *
+   * @return {Promise} A {@link Promise} that will eventually be resovled with a {@link AuthAccess} object
+   */
+  _authenticateResponse(data, continuationCallback) {
+    var authContinuation;
 
-        if (continueData.password) {
-          param.password = continueData.password;
-        }
+    if (data.continuationToken) {
+      // got an AuthContinuation response
+      authContinuation = new AuthContinuation(data);
 
-        return this.transport
-          .post(this.authenticationUrl, this._defaultNonAuthenticatedHeaders(), param);
-      })
-      .then(data => new AuthAccess(data));
+      return continuationCallback(authContinuation)
+        .then(continueData => {
+          if (authContinuation.methods.indexOf(continueData.method) < 0) {
+            throw new Error('The "' + continueData.method + '" authentication protocol is not supported by the server.');
+          }
+          let param = {
+            token: authContinuation.continuationToken,
+            method: continueData.method
+          };
+
+          if (continueData.password) {
+            param.password = continueData.password;
+          }
+
+          return this.transport
+            .post(this.authenticationUrl, this._defaultNonAuthenticatedHeaders(), param);
+        })
+        .then(resp => this._authenticateResponse(resp, continuationCallback));
+    } else if (data.accessToken) {
```

Comment:
> I think that this case should be the first one, do you agree ?

From the order of a regular authentication procedure, one first gets an `continuationToken` reponse.
Actually, it would be better to check the response status code (200 vs. 201) but we don't have access to this so we test properties instead. I don't really see a benefit of changing the order in either way.

> Also, can you add a test when the response contains both (a accessToken and a continuationToken) please ?

Can do. Although that would be considered an invalid JMAP response. Should probably result in an error thrown?